File: mobile/lib/core/models/property_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'property_model.freezed.dart';
part 'property_model.g.dart';

// ignore_for_file: constant_identifier_names

enum PropertyType { MAISON, APPARTEMENT, STUDIO, BUREAU, TERRAIN, BOUTIQUE }

enum PropertyStatus { DISPONIBLE, OCCUPE, RESERVE, MAINTENANCE }

extension PropertyTypeLabel on PropertyType {
  String get label => switch (this) {
        PropertyType.MAISON => 'Maison',
        PropertyType.APPARTEMENT => 'Appartement',
        PropertyType.STUDIO => 'Studio',
        PropertyType.BUREAU => 'Bureau',
        PropertyType.TERRAIN => 'Terrain',
        PropertyType.BOUTIQUE => 'Boutique',
      };
}

extension PropertyStatusLabel on PropertyStatus {
  String get label => switch (this) {
        PropertyStatus.DISPONIBLE => 'Disponible',
        PropertyStatus.OCCUPE => 'Occupé',
        PropertyStatus.RESERVE => 'Réservé',
        PropertyStatus.MAINTENANCE => 'Maintenance',
      };
}

@freezed
class PropertyImageModel with _$PropertyImageModel {
  const factory PropertyImageModel({
    required String id,
    required String url,
    required bool isCover,
    required int order,
  }) = _PropertyImageModel;

  factory PropertyImageModel.fromJson(Map<String, dynamic> json) =>
      _$PropertyImageModelFromJson(json);
}

/// The tenant currently occupying a [PropertyUnitModel], as embedded by
/// `PropertyUnitsMapper.toResponse`'s `currentTenant` (only present/non-null
/// when the unit is OCCUPE and has an active lease).
@freezed
class PropertyUnitCurrentTenantModel with _$PropertyUnitCurrentTenantModel {
  const factory PropertyUnitCurrentTenantModel({
    required String id,
    required String fullName,
    String? phone,
  }) = _PropertyUnitCurrentTenantModel;

  factory PropertyUnitCurrentTenantModel.fromJson(Map<String, dynamic> json) =>
      _$PropertyUnitCurrentTenantModelFromJson(json);
}

/// A single leasable unit within a [PropertyModel] "building". Carries all
/// the fields that used to live directly on Property (status, rent, rooms,
/// surface, ...) - see `PropertyUnitsMapper.toResponse` on the backend.
@freezed
class PropertyUnitModel with _$PropertyUnitModel {
  const factory PropertyUnitModel({
    required String id,
    required String organizationId,
    required String propertyId,
    required String reference,
    String? label,
    String? floor,
    required PropertyType type,
    int? rooms,
    double? surfaceM2,
    required num monthlyRent,
    num? monthlyCharges,
    required PropertyStatus status,
    String? description,
    PropertyUnitCurrentTenantModel? currentTenant,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _PropertyUnitModel;

  factory PropertyUnitModel.fromJson(Map<String, dynamic> json) =>
      _$PropertyUnitModelFromJson(json);
}

extension PropertyUnitModelX on PropertyUnitModel {
  /// Display label: falls back to the reference when no label is set.
  String get displayLabel => label ?? reference;
}

@freezed
class PropertyModel with _$PropertyModel {
  const factory PropertyModel({
    required String id,
    required String organizationId,
    required String reference,
    required String title,
    String? description,
    required PropertyType type,
    required String addressLine,
    required String city,
    String? district,
    double? latitude,
    double? longitude,
    required String ownerId,
    List<PropertyImageModel>? images,
    List<PropertyUnitModel>? units,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _PropertyModel;

  factory PropertyModel.fromJson(Map<String, dynamic> json) =>
      _$PropertyModelFromJson(json);
}

extension PropertyModelX on PropertyModel {
  PropertyImageModel? get coverImage {
    final imgs = images;
    if (imgs == null || imgs.isEmpty) return null;
    return imgs.firstWhere((i) => i.isCover, orElse: () => imgs.first);
  }

  /// Lowest monthly rent among this property's units, or null if it has no
  /// units yet - used for a compact "à partir de ..." summary on list cards.
  num? get minMonthlyRent {
    final u = units;
    if (u == null || u.isEmpty) return null;
    return u.map((unit) => unit.monthlyRent).reduce((a, b) => a < b ? a : b);
  }

  int get availableUnitsCount =>
      units?.where((u) => u.status == PropertyStatus.DISPONIBLE).length ?? 0;
}

File: frontend/src/app/features/properties/models/property.model.ts
import { PaginationQuery } from '../../../core/models';
import { PropertyStatus, PropertyType } from '../../../core/models/enums';

export interface PropertyImage {
  id: string;
  propertyId: string;
  url: string;
  isCover: boolean;
  order: number;
  createdAt: string;
}

export interface PropertyOwnerSummary {
  id: string;
  fullName: string;
  phone: string;
}

export interface PropertyUnitCurrentTenant {
  id: string;
  fullName: string;
  phone: string;
}

/** A single rentable unit within a `Property` — carries all the rental-specific data. */
export interface PropertyUnit {
  id: string;
  organizationId: string;
  propertyId: string;
  reference: string;
  label?: string | null;
  floor?: string | null;
  type: PropertyType;
  rooms?: number | null;
  surfaceM2?: number | null;
  monthlyRent: number;
  monthlyCharges?: number | null;
  status: PropertyStatus;
  description?: string | null;
  currentTenant: PropertyUnitCurrentTenant | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreatePropertyUnitDto {
  reference: string;
  label?: string;
  floor?: string;
  type: PropertyType;
  rooms?: number;
  surfaceM2?: number;
  monthlyRent: number;
  monthlyCharges?: number;
  description?: string;
}

export interface UpdatePropertyUnitDto extends Partial<CreatePropertyUnitDto> {
  status?: PropertyStatus;
}

export interface PropertyUnitFilters {
  type?: PropertyType;
  status?: PropertyStatus;
  minRent?: number;
  maxRent?: number;
  search?: string;
}

export type PropertyUnitQuery = PaginationQuery & PropertyUnitFilters;

/** `Property` is now just the building/listing container — rental details live on `PropertyUnit`. */
export interface Property {
  id: string;
  organizationId: string;
  reference: string;
  title: string;
  description?: string | null;
  type: PropertyType;
  addressLine: string;
  city: string;
  district?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  ownerId: string;
  owner?: PropertyOwnerSummary;
  images?: PropertyImage[];
  units?: PropertyUnit[];
  createdAt: string;
  updatedAt: string;
}

export interface PropertyHistoryEntry {
  id: string;
  propertyId: string;
  changedById: string;
  field: string;
  oldValue?: string | null;
  newValue?: string | null;
  createdAt: string;
}

export interface CreatePropertyDto {
  title: string;
  description?: string;
  type: PropertyType;
  addressLine: string;
  city: string;
  district?: string;
  latitude?: number;
  longitude?: number;
  ownerId: string;
}

export type UpdatePropertyDto = Partial<CreatePropertyDto>;

export interface PropertyFilters {
  type?: PropertyType;
  city?: string;
  search?: string;
  /** Only returns properties having at least one unit in this status (relational filter). */
  unitStatus?: PropertyStatus;
  /** Only returns properties having at least one unit with a rent >= this value. */
  minRent?: number;
  /** Only returns properties having at least one unit with a rent <= this value. */
  maxRent?: number;
}

export type PropertyQuery = PaginationQuery & PropertyFilters;
